research/picai/nnunet_scripts/eval.py

```diff
@@ -1,26 +1,188 @@
+from os.path import join
+from typing import Any, Dict, Iterable, List, Optional
 
-from picai_eval import evaluate_folder
+import numpy as np
+from numpy.typing import NDArray
+
+with contextlib.redirect_stdout(open(os.devnull, "w")):
+    from picai_eval import evaluate, evaluate_folder
+    from picai_eval.image_utils import read_image
+    from picai_eval.metrics import Metrics as PicaiEvalMetrics
+    from report_guided_annotation import extract_lesion_candidates
+
+warnings.simplefilter("ignore", category=FutureWarning)
+
+
+def load_images_from_folder(
+    folder: str,
+    case_identifiers: List[str],
+    postfixes: Optional[List[str]] = None,
+    extensions: List[str] = [".nii.gz", ".nii", ".mha", ".mhd", ".npz", ".npy"],
+) -> NDArray:
+    """
+    Loads images from a folder given a list of case identifiers
+
+    Args:
+        folder (str): The folder containing the images
+        case_identifiers (List[str]): A list of case identifiers for each
+            file. Typically just the filenames without the extension
+        postfixes (Optional[List[str]], optional): A list of strings to append
+            to the case identifiers when looking for files. For example
+            '_labels'. Defaults to None.
+        extensions (List[str], optional): A list of possible image extensions.
+            Defaults to [".nii.gz", ".nii", ".mha", ".mhd", ".npz", ".npy"].
+
+    Returns:
+        NDArray: A numpy array containing the images for each case identifier.
+            The first dimension will be the number of images and thos images
+            will be in the same order as was given by the case_identifiers
+            argument
+    """
+    if postfixes is None:
+        postfixes = [""]
+
+    images = []
+    for case in case_identifiers:
+        for postfix in postfixes:
+            for ext in extensions:
+                path = join(folder, case + postfix + ext)
+                if os.path.exists(path):
+                    images.append(read_image(path))
+
+    return np.stack(images)
+
+
+def get_detection_maps(probability_maps: NDArray) -> NDArray:
+    """
+    Generates detection maps from probability maps by doing lesion extraction
+
+    Args:
+        probability_maps (NDArray): A numpy array containing the predicted
+            probability maps. Should be shape (num_samples, num_classes, ...)
+
+    Returns:
+        NDArray: A numpy array of the detection maps for each class. Note that
+            a detection map is not created for the background class which is
+            assumed to be at index 0. Therefore the output shape is
+            (num_samples, num_classes-1)
+    """
+    num_classes = probability_maps.shape[1]
+
+    detection_maps = []
+    for prob_map in probability_maps:
+        cls_probs = []
+        for cls in range(1, num_classes):  # Ignore background class at idx 0
+            cls_probs.append(extract_lesion_candidates(prob_map[cls], num_lesions_to_extract=5)[0])
```

Comment:
Maybe worth elaborating why we are setting num_lesions_to_extract to 5. If it is an arbitrary parameter that we may want to change, perhaps we include as an argument to this function and default to 5.

Reply:
Leaving as is as this function got moved to the legacy folder, but addressed in the new file by passing kwargs